# Data Understanding

## Objective

Understand the dataset before performing any transformation or analysis.

The purpose of this notebook is to identify:

- Tables
- Relationships
- Keys
- Business meaning
- Data quality issues

## Source Files

- olist_customers_dataset.csv
- olist_geolocation_dataset.csv
- olist_order_items_dataset.csv
- olist_order_payments_dataset.csv
- olist_order_reviews_dataset.csv
- olist_orders_dataset.csv
- olist_products_dataset.csv
- olist_sellers_dataset.csv
- product_category_name_translation.csv

In [1]:
from pathlib import Path

import pandas as pd

In [2]:
PROJECT_ROOT = Path.cwd().parent

RAW_DATA_PATH = PROJECT_ROOT / "data" / "raw"

In [3]:
print(PROJECT_ROOT)
print(RAW_DATA_PATH)

c:\Users\Guilherme Bicudo\Desktop\Olist-Business-Intelligence
c:\Users\Guilherme Bicudo\Desktop\Olist-Business-Intelligence\data\raw


In [4]:
csv_files = list(RAW_DATA_PATH.glob("*.csv"))

for file in csv_files:
    print(file.name)

olist_customers_dataset.csv
olist_geolocation_dataset.csv
olist_orders_dataset.csv
olist_order_items_dataset.csv
olist_order_payments_dataset.csv
olist_order_reviews_dataset.csv
olist_products_dataset.csv
olist_sellers_dataset.csv
product_category_name_translation.csv


In [5]:
for file in csv_files:
    table_name = file.stem.replace("olist_", "").replace("_dataset", "")
    print(table_name)

customers
geolocation
orders
order_items
order_payments
order_reviews
products
sellers
product_category_name_translation


In [6]:
datasets = {}

for file in csv_files:
    table_name = file.stem.replace("olist_", "").replace("_dataset", "")
    datasets[table_name] = pd.read_csv(file)

In [7]:
datasets.keys()

datasets["orders"].head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


## Dataset Summary

In [8]:
overview = []

for name, df in datasets.items():
    overview.append(
        {
            "Dataset": name,
            "Rows": df.shape[0],
            "Columns": df.shape[1],
            "Missing Values": df.isnull().sum().sum(),
            "Duplicate Rows": df.duplicated().sum(),
        })
    
overview = pd.DataFrame(overview)

overview

,Dataset,Rows,Columns,Missing Values,Duplicate Rows
0,customers,99441,5,0,0
1,geolocation,1000163,5,0,261831
2,orders,99441,8,4908,0
3,order_items,112650,7,0,0
4,order_payments,103886,5,0,0
5,order_reviews,99224,7,145903,0
6,products,32951,9,2448,0
7,sellers,3095,4,0,0
8,product_category_name_translation,71,2,0,0


## Orders Dataset Exploration

In this section, we explore the structure of the `orders` table, identifying its columns, data types, and general characteristics before performing any transformations.

In [9]:
orders = datasets["orders"]

In [10]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [11]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')

In [12]:
orders.dtypes

order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object

In [13]:
orders["order_status"].unique()

<StringArray>
[  'delivered',    'invoiced',     'shipped',  'processing', 'unavailable',
    'canceled',     'created',    'approved']
Length: 8, dtype: str

In [14]:
orders["order_status"].nunique()

8

In [15]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

### Key Findings

- The `order_status` column contains **8 unique categories**, indicating a low-cardinality categorical variable.
- Most orders are marked as `delivered`, suggesting a high order fulfillment rate.
- A small proportion of orders were canceled or became unavailable.
- Only a few orders remain in the initial stages (`created` and `approved`), which is expected because this dataset represents historical data.

### Candidate Primary Key Validation

In [16]:
rows = orders.shape[0]
unique_ids = orders["order_id"].nunique()
null_ids = orders["order_id"].isna().sum()

print(f"Rows: {rows}")
print(f"Unique IDs: {unique_ids}")
print(f"Null IDs: {null_ids}")

Rows: 99441
Unique IDs: 99441
Null IDs: 0


#### Key Findings

- `order_id` contains no missing values.
- The number of unique `order_id` values matches the total number of rows.
- Therefore, `order_id` is a strong candidate for the primary key of the `orders` table.

### Dataset Information

This section provides a general overview of the `orders` dataset, including the number of rows, columns, data types, non-null values, and memory usage.

In [17]:
orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 6.1 MB


In [18]:
orders.describe(include="all")

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
count,99441,99441,99441,99441,99281,97658,96476,99441
unique,99441,99441,8,98875,90733,81018,95664,459
top,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2018-03-31 15:08:21,2018-02-27 04:31:10,2018-05-09 15:48:00,2018-05-14 20:02:44,2017-12-20 00:00:00
freq,1,1,96478,3,9,47,3,522


## Data Dictionary

This section documents the columns of the `orders` table, providing a clear description of each field, its data type, nullability, and example values. This documentation improves data understanding and serves as a reference for future ETL, data modeling, and dashboard development.

| Column | Description | Data Type | Nullable | Example |
|--------|-------------|-----------|----------|----------|
| order_id | Unique identifier for each order | String | No | e481f51cbdc54678b7cc49136f2d6af7 |
| customer_id | Unique identifier for each customer | String | No | 9ef432eb6251297304e76186b10a928d |
| order_status | Current status of the order | String | No | delivered |
| order_purchase_timestamp | Timestamp when the customer placed the order | Datetime | No | 2018-03-31 15:08:21 |
| order_approved_at | Timestamp when the order payment was approved | Datetime | Yes | 2018-02-27 04:31:10 |
| order_delivered_carrier_date | Timestamp when the order was handed over to the carrier | Datetime | Yes | 2018-05-09 15:48:00 |
| order_delivered_customer_date | Timestamp when the order was delivered to the customer | Datetime | Yes | 2018-05-14 20:02:44 |
| order_estimated_delivery_date | Estimated delivery date provided to the customer | Datetime | No | 2017-12-20 00:00:00 |

## Customers Dataset Exploration

This section explores the structure and characteristics of the `customers` table, identifying its columns, data types, keys, and relationships with other datasets.

In [19]:
customers = datasets["customers"]

In [20]:
customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [21]:
customers.columns

Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='str')

In [22]:
customers.dtypes

customer_id                   str
customer_unique_id            str
customer_zip_code_prefix    int64
customer_city                 str
customer_state                str
dtype: object

In [23]:
customers["customer_id"].nunique()

99441

In [24]:
customers["customer_unique_id"].nunique()

96096

In [25]:
customers.shape[0]

99441

In [26]:
customers["customer_unique_id"].value_counts().head()

customer_unique_id
8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
6469f99c1f9dfae7733b25662e7f1782     7
ca77025e7201e3b30c44b472ff346268     7
Name: count, dtype: int64

In [27]:
customers.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   customer_id               99441 non-null  str  
 1   customer_unique_id        99441 non-null  str  
 2   customer_zip_code_prefix  99441 non-null  int64
 3   customer_city             99441 non-null  str  
 4   customer_state            99441 non-null  str  
dtypes: int64(1), str(4)
memory usage: 3.8 MB


In [28]:
customers.isna().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

### Customer Dataset Quality Assessment

The `customers` table contains 99,441 records and 5 columns.

The dataset does not contain missing values, meaning all customer attributes are populated.

The analysis identified:

- `customer_id` as a candidate primary key, since it contains 99,441 unique values matching the total number of records.
- `customer_unique_id` as a customer-level identifier, containing 96,096 unique values, indicating that some customers have multiple records in the dataset.

This distinction is important because `customer_id` represents the customer record associated with an order, while `customer_unique_id` represents the actual customer entity.

### Key Findings

- The `customers` table contains 99,441 records and no missing values.
- `customer_id` can be considered a candidate primary key.
- `customer_unique_id` should not be treated as a primary key because multiple records can belong to the same customer.
- The relationship between customers and orders is expected to be one-to-many (`customers` → `orders`).

### Relationship with Orders

The `customers` table is expected to have a one-to-many relationship with the `orders` table.

Relationship:

customers.customer_id → orders.customer_id

One customer can be associated with multiple orders.

## Order Items Dataset Exploration

This section explores the structure of the `order_items` table, identifying its columns, data types, keys, and relationship with orders and products.

In [29]:
order_items = datasets["order_items"]

In [30]:
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [31]:
order_items.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='str')

In [32]:
order_items.dtypes

order_id                   str
order_item_id            int64
product_id                 str
seller_id                  str
shipping_limit_date        str
price                  float64
freight_value          float64
dtype: object

In [33]:
order_items.shape[0]

112650

In [34]:
order_items["order_id"].nunique()

98666

In [35]:
order_items["order_item_id"].nunique()

21

In [36]:
order_items.duplicated(
    subset=["order_id", "order_item_id"]
).sum()

np.int64(0)

### Key Validation

The `order_items` table does not contain a single column that uniquely identifies each record.

The analysis identified a composite key candidate:

`order_id` + `order_item_id`

Validation results:

- Total rows: 112,650
- Unique `order_id`: 98,666
- Unique `order_item_id`: 21
- Duplicate combinations of (`order_id`, `order_item_id`): 0

Therefore, the combination of `order_id` and `order_item_id` can be considered a candidate composite primary key for the table.

In [37]:
order_items.describe()

,order_item_id,price,freight_value
count,112650.000000,112650.000000,112650.000000
mean,1.197834,120.653739,19.990320
std,0.705124,183.633928,15.806405
min,1.000000,0.850000,0.000000
25%,1.000000,39.900000,13.080000
50%,1.000000,74.990000,16.260000
75%,1.000000,134.900000,21.150000
max,21.000000,6735.000000,409.680000


In [38]:
order_items[["price", "freight_value"]].describe()

,price,freight_value
count,112650.000000,112650.000000
mean,120.653739,19.990320
std,183.633928,15.806405
min,0.850000,0.000000
25%,39.900000,13.080000
50%,74.990000,16.260000
75%,134.900000,21.150000
max,6735.000000,409.680000


### Price and Freight Analysis

The `order_items` table contains 112,650 records with financial information related to each purchased item.

#### Price

- Average item price: 120.65
- Median item price: 74.99
- Minimum price: 0.85
- Maximum price: 6,735.00

The difference between the mean and median indicates a right-skewed distribution, suggesting the presence of high-value products affecting the average.

#### Freight Value

- Average freight value: 19.99
- Median freight value: 16.26
- Minimum freight value: 0.00
- Maximum freight value: 409.68

Both price and freight values contain potential outliers that should be investigated before any transformation or removal.

## Products Dataset Exploration

In [39]:
products = datasets["products"]

products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [40]:
products.columns

Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='str')

In [41]:
products.dtypes

product_id                        str
product_category_name             str
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object

In [42]:
products.shape[0]

32951

In [43]:
products["product_id"].nunique()

32951

In [44]:
products["product_id"].isna().sum()

np.int64(0)

In [45]:
products.isna().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

### Products Dataset Quality Assessment

The `products` table contains 32,951 records and represents the product dimension.

The analysis identified:

- `product_id` as a candidate primary key, containing 32,951 unique values matching the total number of records.
- No missing values were found in the `product_id` column.

Missing values were identified mainly in product descriptive attributes:

- `product_category_name`: 610 missing values
- `product_name_length`: 610 missing values
- `product_description_length`: 610 missing values
- `product_photos_qty`: 610 missing values

Physical attributes contain only 2 missing values:

- `product_weight_g`
- `product_length_cm`
- `product_height_cm`
- `product_width_cm`

These missing values should be considered during the data preparation phase, especially because physical attributes can impact logistics and freight analysis.

In [46]:
products["product_category_name"].nunique()

73

In [47]:
products["product_category_name"].value_counts().head(10)

product_category_name
cama_mesa_banho           3029
esporte_lazer             2867
moveis_decoracao          2657
beleza_saude              2444
utilidades_domesticas     2335
automotivo                1900
informatica_acessorios    1639
brinquedos                1411
relogios_presentes        1329
telefonia                 1134
Name: count, dtype: int64

In [48]:
products["product_category_name"].isna().sum()

np.int64(610)

### Product Category Analysis

The `product_category_name` column contains 73 different product categories.

The most frequent categories are:

- cama_mesa_banho: 3,029 products
- esporte_lazer: 2,867 products
- moveis_decoracao: 2,657 products
- beleza_saude: 2,444 products
- utilidades_domesticas: 2,335 products

The category distribution indicates a diverse product catalog, with a strong presence of home, personal care, and leisure-related products.

A total of 610 products do not have a category assigned, representing approximately 1.85% of the product catalog. These records should be evaluated during the data preparation phase.

## Order Payments Dataset Exploration

In [49]:
order_payments = datasets["order_payments"]

order_payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [50]:
order_payments.columns

Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='str')

In [51]:
order_payments.dtypes

order_id                    str
payment_sequential        int64
payment_type                str
payment_installments      int64
payment_value           float64
dtype: object

In [52]:
order_payments.shape[0]

103886

In [53]:
order_payments["order_id"].nunique()

99440

In [54]:
order_payments["payment_sequential"].nunique()

29

In [55]:
order_payments.duplicated(
    subset=["order_id", "payment_sequential"]
).sum()

np.int64(0)

In [56]:
order_payments["payment_type"].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64

### Order Payments Dataset Quality Assessment

The `order_payments` table contains 103,886 payment records associated with customer orders.

The analysis identified:

- `order_id` as a foreign key connecting payments with the `orders` table.
- `payment_sequential` as a payment sequence identifier within each order.
- The combination of `order_id` and `payment_sequential` was validated as a candidate composite key, with no duplicate combinations found.

The dataset contains 99,440 unique orders, indicating that some orders have multiple payment records.

Payment methods distribution:

- Credit card: 76,795 records
- Boleto: 19,784 records
- Voucher: 5,775 records
- Debit card: 1,529 records
- Not defined: 3 records

Credit card is the dominant payment method, representing approximately 74% of all payment records.

## Sellers Dataset Exploration

In [57]:
sellers = datasets["sellers"]

sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [58]:
sellers.columns

Index(['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state'], dtype='str')

In [59]:
sellers.dtypes

seller_id                   str
seller_zip_code_prefix    int64
seller_city                 str
seller_state                str
dtype: object

In [60]:
sellers.shape[0]

3095

In [61]:
sellers["seller_id"].nunique()

3095

In [62]:
sellers["seller_id"].isna().sum()

np.int64(0)

In [63]:
sellers.isna().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

### Key Findings

- The `seller_id` column contains 3,095 unique values, matching the total number of rows.
- No missing values were identified in the `seller_id` column, making it a strong candidate for the primary key of the `sellers` table.
- The `sellers` table contains geographic information through `seller_city`, `seller_state`, and `seller_zip_code_prefix`, enabling regional analysis of seller activity.
- The table can be related to the `order_items` table through the `seller_id` column.

# Order Reviews Dataset Exploration

This section explores the structure of the order_reviews table, identifying its columns, data types, key characteristics, customer satisfaction metrics, and relationship with the orders table.

In [64]:
order_reviews = datasets["order_reviews"]

order_reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [65]:
order_reviews.columns

Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='str')

In [66]:
order_reviews.dtypes

review_id                    str
order_id                     str
review_score               int64
review_comment_title         str
review_comment_message       str
review_creation_date         str
review_answer_timestamp      str
dtype: object

In [67]:
order_reviews.shape[0]

99224

In [68]:
order_reviews["review_id"].nunique()

98410

In [69]:
order_reviews["order_id"].nunique()

98673

In [70]:
order_reviews.duplicated(
    subset=["review_id"]
).sum()

np.int64(814)

In [71]:
order_reviews["review_score"].value_counts()

review_score
5    57328
4    19142
1    11424
3     8179
2     3151
Name: count, dtype: int64

In [72]:
order_reviews.isna().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

## Key Findings

- The `review_id` column contains 98,410 unique values, which is lower than the total number of rows, indicating that it cannot be considered a strict primary key.
- The `order_id` column also contains repeated values, showing that multiple reviews can be associated with the same order.
- The `review_score` distribution shows that most customers provided positive evaluations, with score 5 being the most frequent category.
- Text fields such as `review_comment_title` and `review_comment_message` contain a high number of missing values, which is expected since customers are not required to provide written feedback.
- The table can be related to the `orders` table through the `order_id` column.

# Data Model Overview

After exploring the available datasets, the main entities, relationships, and table granularities were identified.

This overview represents the logical structure of the dataset before data transformation and analytical modeling.

## Dataset Structure

| Table | Grain | Primary Key / Candidate Key | Main Relationships |
|---|---|---|---|
| orders | 1 row = 1 order | order_id | customer_id, order_items, order_payments, order_reviews |
| customers | 1 row = 1 customer record | customer_id | orders |
| order_items | 1 row = 1 product item within an order | order_id + order_item_id | orders, products, sellers |
| products | 1 row = 1 product | product_id | order_items |
| sellers | 1 row = 1 seller | seller_id | order_items |
| order_payments | 1 row = 1 payment transaction | order_id + payment_sequential | orders |
| order_reviews | 1 row = 1 customer review event | review_id (not unique) | orders |
| geolocation | 1 row = 1 postal code location | zip_code_prefix (candidate) | customers, sellers |
| product_category_name_translation | 1 row = 1 category translation | product_category_name | products |

## Logical Relationships


                         customers
                             |
                             |
                         customer_id
                             |
                             ↓
                         orders
                             |
              +--------------+--------------+
              |              |              |
              ↓              ↓              ↓
        order_items    order_payments  order_reviews
              |
        +-----+------+
        |            |
        ↓            ↓
    products      sellers

## Key Modeling Observations

- The `orders` table represents the central entity of the business process, containing one record per order.
- The `order_items` table represents the transactional level, containing the products purchased in each order.
- The `customers`, `products`, and `sellers` tables represent descriptive entities that can be modeled as dimensions.
- The `order_payments` and `order_reviews` tables represent events associated with orders and may contain multiple records per order.
- The dataset structure suggests a dimensional model approach for analytical purposes, with `order_items` as the main fact table.